dango/lending-pool/src/execute.rs

```diff
@@ -0,0 +1,237 @@
+    // Ensure margin accounts can't deposit
+    ensure_sender_account_is_not_margin(&ctx)?;
+
+    // For each deposited denom, ensure it's whitelisted and mint LP tokens.
+    let cfg = ctx.querier.query_config()?;
+    let mut msgs = vec![];
+    for coin in ctx.funds.into_iter() {
+        ensure!(
+            WHITELISTED_DENOMS.has(ctx.storage, coin.denom.clone()),
+            "Invalid denom"
+        );
+
+        let lp_denom = Denom::from_parts(vec![
+            Part::from_str(NAMESPACE)?,
+            Part::from_str("lp")?,
+            Part::from_str(&coin.denom.to_string())?,
+        ])?;
+
+        msgs.push(Message::execute(
+            cfg.bank,
+            &bank::ExecuteMsg::Mint {
+                to: recipient.unwrap_or(ctx.sender),
+                denom: lp_denom,
+                amount: coin.amount,
+            },
+            Coins::new(),
+        )?);
+    }
+
+    Ok(Response::new().add_messages(msgs))
+}
+
+pub fn withdraw(ctx: MutableCtx, recipient: Option<Addr>) -> anyhow::Result<Response> {
+    // Ensure margin accounts can't withdraw
+    ensure_sender_account_is_not_margin(&ctx)?;
+
+    // Unwrap the recipient
+    let recipient = recipient.unwrap_or(ctx.sender);
+
+    // Ensure there are funds to withdraw
+    ensure!(!ctx.funds.is_empty(), "No funds to withdraw");
+
+    let cfg = ctx.querier.query_config()?;
+    let mut msgs = vec![];
+    let mut withdrawn = Coins::new();
+    for coin in ctx.funds.into_iter() {
+        // Ensure only LP tokens are sent
+        ensure!(
+            coin.denom.inner().len() == 3
+                && coin.denom.namespace().is_some_and(|ns| **ns == NAMESPACE)
+                && coin.denom.inner().to_vec()[1].as_str() == "lp",
+            "Invalid denom"
+        );
+
+        // Add msg to send the underlying tokens to the recipient
+        let underlying_denom = Denom::from_parts(vec![coin.denom.inner()[2].clone()])?;
+        let amount = coin.amount;
+        withdrawn.insert(Coin::new(underlying_denom, amount)?)?;
+
+        // Burn the LP tokens
+        msgs.push(Message::execute(
+            cfg.bank,
+            &bank::ExecuteMsg::Burn {
+                from: ctx.contract,
+                denom: coin.denom,
+                amount,
+            },
+            Coins::new(),
+        )?);
+    }
+
+    // Transfer the underlying tokens to the recipient
+    msgs.push(Message::transfer(recipient, withdrawn)?);
+
+    Ok(Response::new().add_messages(msgs))
+}
+
+pub fn borrow(ctx: MutableCtx, coins: Coins) -> anyhow::Result<Response> {
+    // Ensure sender is a margin account
+    ensure_sender_account_is_margin(&ctx)?;
+
+    // Ensure the coins are whitelisted and lending pool has enough liquidity
+    for coin in coins.clone().into_iter() {
+        ensure!(
+            WHITELISTED_DENOMS.has(ctx.storage, coin.denom.clone()),
+            "Invalid denom. Only whitelisted denoms can be borrowed."
+        );
+    }
```

Comment:
clone is not needed here
```suggestion
    for coin in &coins {
        ensure!(
            WHITELISTED_DENOMS.has(ctx.storage, coin.denom.clone()),
            "Invalid denom. Only whitelisted denoms can be borrowed."
        );
    }
```